src/Oscoin/Node/Mempool.hs

```diff
@@ -31,39 +31,36 @@ import qualified Data.Foldable as Fold
 
 -- | Handle to a mutable 'Mempool'.
 data Handle c tx = Handle
-    { hMempool   :: TVar (Mempool c tx)
-    , hBroadcast :: Channel tx
+    { hMempool    :: TVar (Mempool c tx)
+    , hBroadcast  :: Channel tx
+    , hTxValidate :: tx -> Either (TxValidationError c tx) ()
     }
 
 -- | Create a new 'Handle' with an underlying empty mempool.
-new :: Ord (Hash c) => STM (Handle c tx)
-new = do
+new :: Ord (Hash c) => TxValidator c tx -> STM (Handle c tx)
+new hTxValidate = do
     hMempool   <- STM.newTVar mempty
     hBroadcast <- STM.newBroadcastTChan
     pure Handle{..}
 
 -- | Like 'new', but without running a transaction.
-newIO :: Ord (Hash c) => IO (Handle c tx)
-newIO = do
+newIO :: Ord (Hash c) => TxValidator c tx -> IO (Handle c tx)
+newIO hTxValidate = do
     hMempool   <- STM.newTVarIO mempty
     hBroadcast <- STM.newBroadcastTChanIO
     pure Handle{..}
 
-insert :: (Crypto.Hashable c tx) => Handle c tx -> tx -> STM ()
-insert Handle{hMempool, hBroadcast} tx = do
-    STM.modifyTVar' hMempool (Internal.insert tx)
-    STM.writeTChan hBroadcast (Insert [tx])
-
-insertMany
+insert
     :: ( Crypto.Hashable c tx
-       , Foldable t
        )
     => Handle c tx
-    -> t tx
-    -> STM ()
-insertMany Handle{hMempool, hBroadcast} txs = do
-    STM.modifyTVar' hMempool (Internal.insertMany txs)
-    STM.writeTChan hBroadcast (Insert (Fold.toList txs))
+    -> tx
+    -> STM (Either (TxValidationError c tx) ())
+insert Handle{hMempool, hBroadcast, hTxValidate} tx = runExceptT $ do
+    ExceptT $ pure $ hTxValidate tx
+    lift $ STM.modifyTVar' hMempool (Internal.insert tx)
+    lift $ STM.writeTChan hBroadcast (Insert [tx])
```

Comment:
Since only `hTxValidate` returns an Either here, wouldn't it be better to just `case` on that, and remove the `runExceptT`` machinery and lifts?

Reply:
https://github.com/oscoin/oscoin/pull/555/commits/03cd7f8d81783a06a3a24bf3b90ffea3fb75fcb7